### PR TITLE
more debug logging and warning for missing data.context

### PR DESCRIPTION
**Pull request by jqian2017**

### CODE REVIEW

File: middleware/index.ts [hunk 1]
```diff
@@ -151,12 +151,12 @@ function createOrg(logger: Logger, reqContext: any, accessToken?: string): Org {
  * @param functionInvocationId -- FunctionInvocationRequest ID, if applicable
  * @return context
  */
-function createContext(id: string, logger: Logger, reqContext: any, accessToken?: string, functionInvocationId?: string): Context {
-    if (typeof reqContext === 'string') {
+function createContext(id: string, logger: Logger, reqContext?: any, accessToken?: string, functionInvocationId?: string): Context {
+    if (reqContext && typeof reqContext === 'string') {
```
Comment (cwallsfdc) [inline middleware/index.ts]: When would `reqContent` be undefined?

File: middleware/index.ts [hunk 2]
```diff
@@ -168,6 +168,22 @@ function createContext(id: string, logger: Logger, reqContext: any, accessToken?
...
+/**
+ * TODO: replace this with salesforce-sdk api call to get secret
+ * @param name
+ * @param key
+ * @param logger
+ */
+function getSecret(name, key, logger: Logger) : string {
+    try {
+        let data = fs.readFileSync(`/platform/services/${name}/secret/${key}`);
+        return data.toString();
```
Comment (cwallsfdc) [inline middleware/index.ts]: Could `data` be null/undefined?

File: middleware/index.ts [hunk 3]
```diff
@@ -1,4 +1,5 @@
-
+import * as fs from 'fs';
+import {LoggerLevel} from '@salesforce/core'
```
Comment (dhagberg-sf) [inline middleware/index.ts]: nit: inconsistent trailing `;` chars - I think most of this source *does* use semicolons?

File: middleware/index.ts (hunk 1 above)
Comment (dhagberg-sf) [inline middleware/index.ts]: should be safe to only check `typeof reqContext` even if `undefined` -- typeof should return `undefined`.

File: middleware/index.ts [hunk 4]
```diff
@@ -168,6 +168,22 @@ function createContext(id: string, logger: Logger, reqContext: any, accessToken?
...
+ * @param name
+ * @param key
+ * @param logger
+ */
+function getSecret(name, key, logger: Logger) : string {
+    try {
+        let data = fs.readFileSync(`/platform/services/${name}/secret/${key}`);
+        return data.toString();
+    } catch(err) {
+        logger.info(`Failed to read secret name "${name}" key "${key}": ${err.stack}`);
```
Comment (dhagberg-sf) [inline middleware/index.ts]: probably `logger.debug` depending on how often we're calling it.

File: middleware/index.ts [hunk 5]
```diff
@@ -184,13 +200,29 @@ export default function applySfFxMiddleware(request: any, state: any, resultArgs
         throw new Error('Request Data not provided');
     }
 
+    // Logger is the last element in the array
+    const logger: Logger = resultArgs[resultArgs.length-1];
+    if (!logger) {
+        throw new Error('Logger not provided in resultArgs from node system function');
+    }
+
+    //use secret her in lieu of DEBUG runtime environment var until we haave deployment time support of config var
```
Comment (cwallsfdc) [inline middleware/index.ts]: her => here.
haave => have.

File: middleware/index.ts [hunk 6]
```diff
@@ -184,13 +200,29 @@ export default function applySfFxMiddleware(request: any, state: any, resultArgs
...
+
     const data = request.payload.data;
     if (!data) {
-        throw new Error('Data not provided');
+        throw new Error('Data field of the cloudEvent not provided in the request');
     }
 
     if (!data.context) {
-        throw new Error('Context not provided in data');
+        logger.warn('Context not provided in data, context.org is not initialized');
```
Comment (cwallsfdc) [inline middleware/index.ts]: It's actually more than `context.org`.  Can we say ": context partially initialize"?

File: middleware/index.ts [hunk 7]
```diff
@@ -184,13 +200,29 @@ export default function applySfFxMiddleware(request: any, state: any, resultArgs
...
+
+    //use secret her in lieu of DEBUG runtime environment var until we haave deployment time support of config var
+    const debugSecret = getSecret("sf-debug", "DEBUG", logger);
+    logger.info(`DEBUG flag is ${debugSecret}`);
+    if(debugSecret || LoggerLevel.DEBUG === logger.getLevel() || process.env.DEBUG) {
+        //for dev preview, we log the ENTIRE raw request, may need to filter sensitive properties out later
+        //the hard part of filtering to know which name to filter
+        let childLogger: Logger = logger.child('raw_request', request);
+        childLogger.info('debug raw request in middleware');
+    }
```
Comment (cwallsfdc) [inline middleware/index.ts]: Got test coverage for `getSecret` and above?

File: middleware/index.ts (hunk 4 above)
Comment (cwallsfdc) [reply middleware/index.ts]: Also, suggest not logging `stack`.

Comment (jqian2017) [reply middleware/index.ts]: @dhagberg-sf , logger.debug will never log anything in a deployed environment, as there is no way to set that DEBUG=1 runtime env in deployed evergreen...hence info

File: middleware/index.ts [hunk 8]
```diff
@@ -109,7 +109,7 @@ function createUser(userContext: any): User {
  * @param reqContext
  * @return org
  */
-function createOrg(logger: Logger, reqContext: any, accessToken?: string): Org {
+function createorg(logger: Logger, reqContext: any, accessToken?: string): Org {
```
[the remark this answers is not in the document]
Comment (jqian2017) [reply middleware/index.ts]: must be typo

File: middleware/index.ts (hunk 2 above)
Comment (dhagberg-sf) [reply middleware/index.ts]: I think it could be and empty byte array, should never be null/undefined (readFileSync would throw an error in that case).

Comment (jqian2017) [reply middleware/index.ts]: I'm hoping to replace this with salesforce sdk, so I don't have to worry about this.
But yes... I should add some null checking...
JSYK, from CLI, when you create a secret key, they do not allow you to have secretkey=, you must have something on the RHS of the =

File: middleware/index.ts (hunk 3 above)
Comment (jqian2017) [reply middleware/index.ts]: I should add some linting to this project...

File: middleware/index.ts (hunk 1 above)
Comment (jqian2017) [reply middleware/index.ts]: when user does NOT have data.context... which we now allow and only warning user

File: middleware/index.ts (hunk 5 above)
Comment (jqian2017) [reply middleware/index.ts]: good catch... errhh... blame on this macbook keyboard!

File: middleware/index.ts (hunk 7 above)
Comment (jqian2017) [reply middleware/index.ts]: will add for above... for getSecret, will replace with sdk api.. so no plan to test that.

**Comment by jqian2017** [comment]
Here is some sample logging when the secret is set:

`{"name":"Evergreen Logger","hostname":"cat-r3fsfrdlxik-n8fjd-deployment-79fb7b5549-d5gfr","pid":1,"request_id":"00D1F000000HGeX-4UjewoKL-97pGu-qbvXjc-","level":30,"msg":"DEBUG flag is 1","time":"2020-03-13T16:47:05.143Z","v":0}
{"name":"Evergreen Logger","hostname":"cat-r3fsfrdlxik-n8fjd-deployment-79fb7b5549-d5gfr","pid":1,"request_id":"00D1F000000HGeX-4UjewoKL-97pGu-qbvXjc-","level":20,"msg":"debug raw request in middleware","time":"2020-03-13T16:47:05.143Z","v":0}
{"name":"Evergreen Logger","hostname":"cat-r3fsfrdlxik-n8fjd-deployment-79fb7b5549-d5gfr","pid":1,"request_id":"00D1F000000HGeX-4UjewoKL-97pGu-qbvXjc-","level":20,"payload":{"type":"com.salesforce.function.invoke","source":"urn:event:from:salesforce/1F/224.0/00D1F000000HGeX/apex/unknown","id":"00D1F000000HGeX-4UjewoKL-97pGu-qbvXjc-","time":"2020-03-13T16:47:04.794543Z","schemaURL":null,"contentType":"application/json","data":{"context":{"apiVersion":"48.0","payloadVersion":"0.1","userContext":{"orgId":"00D1F000000HGeX","userId":"0051F00000IL4lY","onBehalfOfUserId":null,"username":"test-f2b0ydfkx9lb@example.com","salesforceBaseUrl":"https://customer-customer-7192-dev-ed.cs90.my.salesforce.com","orgDomainUrl":"https://customer-customer-7192-dev-ed.cs90.my.salesforce.com"}},"payload":{"accountName":"Acct","contactName":"MyContact","opportunityName":"Oppty"},"sfContext":{"accessToken":"<access_token - HIDDEN>","functionInvocationId":null,"functionName":"JQAccountFunction","requestId":"00D1F000000HGeX-4UjewoKL-97pGu-qbvXjc-","resource":"https://cat-r3fsfrdlxik.currant-tswmj93.aqueous-spire-5080.herokuspace.com"}},"specVersion":"0.2"},"headers":{"host":"cat-r3fsfrdlxik.currant-tswmj93.aqueous-spire-5080.herokuspace.com","user-agent":"Apache-HttpClient/4.5.6 (Java/11.0.4.0.101)","content-length":"1072","accept-encoding":"gzip,deflate","authorization":"<Authorization - HIDDEN>","content-type":"application/json; charset=UTF-8","forwarded":"for=10.0.51.62;proto=https","x-b3-sampled":"0","x-b3-spanid":"b55dc716f686cf13","x-b3-traceid":"a1338d3659b68ee8b55dc716f686cf13","x-envoy-decorator-operation":"cat-r3fsfrdlxik-n8fjd.currant-tswmj93.svc.cluster.local:80/*","x-envoy-expected-rq-timeout-ms":"600000","x-envoy-internal":"true","x-forwarded-for":"10.0.51.62, 10.0.37.82","x-forwarded-proto":"https","x-istio-attributes":"ClUKGGRlc3RpbmF0aW9uLnNlcnZpY2UuaG9zdBI5EjdjYXQtcjNmc2ZyZGx4aWstbjhmamQuY3VycmFudC10c3dtajkzLnN2Yy5jbHVzdGVyLmxvY2FsClMKF2Rlc3RpbmF0aW9uLnNlcnZpY2UudWlkEjgSNmlzdGlvOi8vY3VycmFudC10c3dtajkzL3NlcnZpY2VzL2NhdC1yM2ZzZnJkbHhpay1uOGZqZAozChhkZXN0aW5hdGlvbi5zZXJ2aWNlLm5hbWUSFxIVY2F0LXIzZnNmcmRseGlrLW44ZmpkCjIKHWRlc3RpbmF0aW9uLnNlcnZpY2UubmFtZXNwYWNlEhESD2N1cnJhbnQtdHN3bWo5MwpOCgpzb3VyY2UudWlkEkASPmt1YmVybmV0ZXM6Ly9pc3Rpby1pbmdyZXNzZ2F0ZXdheS1iYzU2N2M1NDUtejRrZmwuaXN0aW8tc3lzdGVt","x-request-id":"00D1F000000HGeX-4UjewoKL-97pGu-qbvXjc-"},"msg":"","time":"2020-03-13T16:47:05.144Z","v":0}
{"name":"Evergreen Logger","hostname":"cat-r3fsfrdlxik-n8fjd-deployment-79fb7b5549-d5gfr","pid":1,"request_id":"00D1F000000HGeX-4UjewoKL-97pGu-qbvXjc-","level":30,"msg":"Invoking JQAccountFunction...","time":"2020-03-13T16:47:05.144Z","v":0}
`

File: middleware/index.ts [hunk 9]
```diff
@@ -12,8 +13,7 @@ import {
     SuccessResult,
     UnitOfWork,
     User,
-} from '@salesforce/salesforce-sdk';
-
+} from '@salesforce/salesforce-sdk'
```
Comment (cwallsfdc) [inline middleware/index.ts]: Also needs `;`.

File: middleware/.eslintrc.js [hunk 10]
```diff
@@ -0,0 +1,18 @@
...
+    'plugin:@typescript-eslint/eslint-recommended',
+    'plugin:@typescript-eslint/recommended',
+    'prettier/@typescript-eslint',
+  ],
+  rules: {
+    '@typescript-eslint/interface-name-prefix': 'off',
+    'semi': 'off',
+    '@typescript-eslint/semi': ['error'],
+  }
+};
```
Comment (cwallsfdc) [inline middleware/.eslintrc.js]: :clap:

File: middleware/test/unit/ContextTests.ts [hunk 11]
```diff
@@ -1,13 +1,24 @@
 /* tslint:disable: no-unused-expression */
-import { expect } from 'chai';
-import 'mocha';
-
-import applySfFxMiddleware from '../../index';
-import { ConnectionConfig, Constants, Context, Logger} from '@salesforce/salesforce-sdk';
-import { generateData, generateRawMiddleWareRequest } from './FunctionTestUtils';
+import {LoggerLevel} from '@salesforce/core'
```
Comment (cwallsfdc) [inline middleware/test/unit/ContextTests.ts]: Need `;` here and below.

**Pull request merged**